File: Report-Sorting/QuickSort.cpp
#include <bits/stdc++.h>
using namespace std;

// Partition function for QuickSort
int partition(vector<double>& arr, int left, int right) {
    int pivot = right;
    int stored_pivot = left;

    for (int index = left; index < right; ++index) {
        if (arr[index] < arr[pivot])
        {
            swap(arr[index], arr[stored_pivot]);
            stored_pivot++;
        }
    }
    swap(arr[pivot], arr[stored_pivot]);
    return stored_pivot;
}

// QuickSort function
void QuickSort(vector<double>& arr, int left, int right) {
    if (left >= right)
        return;
    int pi = partition(arr, left, right);
    QuickSort(arr, left, pi - 1);
    QuickSort(arr, pi + 1, right);
}
			
// Checking function to validate sorting
bool Checking(vector<double>& arr, int k) {
    string test_file = "./my_testcase/test" + to_string(k + 1) + ".txt";
    vector<double> sorted_arr;
    ifstream file(test_file);
    double value;
    while (file >> value)
        sorted_arr.push_back(value);
    file.close();
    if (arr == sorted_arr)
        return true;
    else
        return false;
}

int main() {
    vector<double> input_arr;
    for (int i = 9; i >=0 ; i--) {
        input_arr.clear();
        string filename = "./my_dataset/data" + to_string(i + 1) + ".txt";
        ifstream file(filename);
        double value;
        while (file >> value)
            input_arr.push_back(value);
        file.close();
        clock_t start = clock();
        QuickSort(input_arr, 0, input_arr.size() - 1);
        clock_t end = clock();
        cout << "Sort data " << i + 1 << ": Done\n";
        cout << "Time taken: " << (double(end - start) / CLOCKS_PER_SEC) * 1000 << endl;
        if (Checking(input_arr, i))
            cout << "Correct!\n";
        else
            cout << "Wrong!\n";
    }
    return 0;
}

File: Report-Sorting/QuickSort.py
import time
import sys

sys.setrecursionlimit(99999999)

# Partition function for QuickSort
def partition(arr, left, right):
    pivot = right
    stored_pivot = left

    for index in range(left, right):
        if arr[index] < arr[pivot]:
            arr[index], arr[stored_pivot] = arr[stored_pivot], arr[index]
            stored_pivot += 1

    arr[pivot], arr[stored_pivot] = arr[stored_pivot], arr[pivot]
    return stored_pivot

# QuickSort function
def QuickSort(arr, left, right):
    if left >= right:
        return
    pi = partition(arr, left, right)
    QuickSort(arr, left, pi - 1)
    QuickSort(arr, pi + 1, right)

# Checking function to validate sorting
def Checking(arr, k):
    test_file = "./my_testcase/test" + str(k + 1) + ".txt"
    with open(test_file, 'r') as file:
        sorted_arr = [float(line.strip()) for line in file]

    return arr == sorted_arr

def main():
    for i in range(9, -1, -1):
        input_arr = []
        filename = "./my_dataset/data" + str(i + 1) + ".txt"
        with open(filename, 'r') as file:
            input_arr = [float(line.strip()) for line in file]

        start = time.time()
        QuickSort(input_arr, 0, len(input_arr) - 1)
        end = time.time()
        print("Sort data {}: Done".format(i + 1))
        print("Time taken: {:.2f} ms".format((end - start) * 1000))

        if Checking(input_arr, i):
            print("Correct!")
        else:
            print("Wrong!")

if __name__ == "__main__":
    main()
